File: apps/web/lib/shelf_mark_ref.ts
/** 书架划线/想法锚点：SHELF.{bookId}.{sectionId}.p{page}@{start}-{end} */

import { formatShelfCheckinLabel, getShelfRefLabel, isShelfRef, parseShelfRef } from './shelf_checkin';

export type ParsedShelfMarkRef = {
  bookId: string;
  sectionId: string;
  pageIndex: number;
  spanStart?: number;
  spanEnd?: number;
  raw: string;
};

export function buildShelfMarkRef(
  bookId: string,
  sectionId: string,
  pageIndex: number,
  span?: { start: number; end: number } | null,
): string {
  const base = `SHELF.${bookId}.${sectionId}.p${Math.max(0, pageIndex)}`;
  if (span && span.end > span.start) {
    return `${base}@${span.start}-${span.end}`;
  }
  return base;
}

export function parseShelfMarkRef(ref: string | null | undefined): ParsedShelfMarkRef | null {
  if (!ref || !isShelfRef(ref)) return null;
  const trimmed = ref.trim();
  const [base, spanPart] = trimmed.split('@');

  const pageMatch = base.match(/^SHELF\.(.+)\.p(\d+)$/);
  if (pageMatch) {
    const rest = pageMatch[1];
    const pageIndex = Number(pageMatch[2]);
    const dot = rest.lastIndexOf('.');
    if (dot <= 0 || Number.isNaN(pageIndex)) return null;
    const bookId = rest.slice(0, dot);
    const sectionId = rest.slice(dot + 1);
    if (!bookId || !sectionId) return null;
    let spanStart: number | undefined;
    let spanEnd: number | undefined;
    if (spanPart) {
      const [s, e] = spanPart.split('-').map(Number);
      if (!Number.isNaN(s) && !Number.isNaN(e) && e > s) {
        spanStart = s;
        spanEnd = e;
      }
    }
    return { bookId, sectionId, pageIndex, spanStart, spanEnd, raw: trimmed };
  }

  const plain = parseShelfRef(trimmed);
  if (!plain) return null;
  return { ...plain, pageIndex: 0, raw: trimmed };
}

export function isShelfMarkRef(ref: string | null | undefined): boolean {
  return parseShelfMarkRef(ref) != null;
}

export function formatShelfMarkRefLabel(
  ref: string,
  bookTitle?: string,
  sectionTitle?: string,
): string {
  const cached = getShelfRefLabel(ref.split('@')[0] ?? ref);
  if (cached) return cached;

  const p = parseShelfMarkRef(ref);
  if (!p) return ref;
  const base = formatShelfCheckinLabel(bookTitle || '书架', sectionTitle || '');
  if (p.pageIndex > 0) return `${base} · 第 ${p.pageIndex + 1} 页`;
  return base;
}

export function shelfMarkHref(ref: string): string {
  const p = parseShelfMarkRef(ref);
  if (!p) return '/shelf';
  const params = new URLSearchParams();
  params.set('section', p.sectionId);
  if (p.pageIndex > 0) params.set('page', String(p.pageIndex));
  return `/shelf/${encodeURIComponent(p.bookId)}?${params.toString()}`;
}

/** 云同步用：去掉 @ 词组偏移。 */
export function shelfSyncRef(ref: string): string {
  return ref.split('@')[0] ?? ref;
}

File: apps/mobile/lib/features/shelf/shelf_mark_ref.dart
/// 书架划线/想法锚点（对齐 Web shelf_mark_ref.ts）
library;

class ShelfMarkRefParsed {
  const ShelfMarkRefParsed({
    required this.bookId,
    required this.sectionId,
    required this.pageIndex,
    this.spanStart,
    this.spanEnd,
  });

  final String bookId;
  final String sectionId;
  final int pageIndex;
  final int? spanStart;
  final int? spanEnd;
}

String buildShelfMarkRef(
  String bookId,
  String sectionId, {
  int pageIndex = 0,
  int? spanStart,
  int? spanEnd,
}) {
  final base = 'SHELF.$bookId.$sectionId.p${pageIndex.clamp(0, 9999)}';
  if (spanStart != null && spanEnd != null && spanEnd > spanStart) {
    return '$base@$spanStart-$spanEnd';
  }
  return base;
}

ShelfMarkRefParsed? parseShelfMarkRef(String? ref) {
  if (ref == null || !ref.startsWith('SHELF.')) return null;
  var rest = ref.substring(6);
  int? spanStart;
  int? spanEnd;
  final at = rest.lastIndexOf('@');
  if (at >= 0) {
    final span = rest.substring(at + 1);
    rest = rest.substring(0, at);
    final parts = span.split('-');
    if (parts.length == 2) {
      spanStart = int.tryParse(parts[0]);
      spanEnd = int.tryParse(parts[1]);
    }
  }
  var pageIndex = 0;
  final pageMatch = RegExp(r'\.p(\d+)$').firstMatch(rest);
  if (pageMatch != null) {
    pageIndex = int.tryParse(pageMatch.group(1)!) ?? 0;
    rest = rest.substring(0, pageMatch.start);
  }
  final dot = rest.lastIndexOf('.');
  if (dot <= 0 || dot >= rest.length - 1) return null;
  return ShelfMarkRefParsed(
    bookId: rest.substring(0, dot),
    sectionId: rest.substring(dot + 1),
    pageIndex: pageIndex,
    spanStart: spanStart,
    spanEnd: spanEnd,
  );
}

String formatShelfMarkRefLabel(String ref) {
  final p = parseShelfMarkRef(ref);
  if (p == null) return '选中文字';
  if (p.spanStart != null && p.spanEnd != null) {
    return '选中片段 · ${p.spanEnd! - p.spanStart!} 字';
  }
  return '本节笔记';
}

int? findPlainTextSpan(String html, String selected) {
  final text = _stripHtml(html);
  final needle = selected.trim();
  if (needle.isEmpty) return null;
  final start = text.indexOf(needle);
  if (start < 0) return null;
  return start;
}

String _stripHtml(String html) {
  return html
      .replaceAll(RegExp(r'<[^>]+>'), '')
      .replaceAll('&nbsp;', ' ')
      .replaceAll('&amp;', '&')
      .replaceAll('&lt;', '<')
      .replaceAll('&gt;', '>')
      .replaceAll(RegExp(r'\s+'), ' ')
      .trim();
}

String plainTextFromHtml(String html) => _stripHtml(html);
